Performance Enhancements

mwm-deriv opened the pull request:
This changeset improves the performance of `Date::Utility` with respect to object creation.

Objects are now created in about a third (~4μs) of the time used on the current `master` branch (~12μs).  This performance benefit accrues to all `Date::Utility` consumers without any code changes.

The implementations changes are largely:

- Use constant regexps which are only compiled once for faster comparisons
- Remove the use of `Tie::Cache::LRU`, the overhead of which was higher than simply creating new objects
- Switch from Moose to Moo for faster MOP object building

mwm-deriv committed:
Move heavily used regexps to constants.

We lose the ability to interpolate sub-expressions, but this avoids a
recompilation on each evalulation which is approximatiely as expensive
as doing the match itself.

mwm-deriv committed:
Drop LRU cache

FETCH and STORE operations can be surprisingly expensive with at tied
LRU cache. In fact, they are slower than just building a new object.

On the other hand, there may be cached attribute build which would also
improve things for using those objects.

I cannot believe that the hit rate for those will be able to dominate
this storage and retrieval problem.

mwm-deriv committed:
Move from Moose to Moo

The major differences from a consumer perspective are:

- `epoch` is no longer checked for proper type
- Attempting to build an object from another object actually returns a
  new object, not the same one

These are consequences of te diffeence between Moo and Moose.  But they
seem, to me, inconsequential.

mwm-deriv committed:
Improve BUILDARGS flow

- Use early returns to make it obvious what we are doing
- Don't preallocate a hashref which is directly returned
- Reorder based on expected usage
- Reduce checks inside a provided hashref

mwm-deriv committed:
Alter match order

This organization reflects ease of matching (shorter is better) and
liklelihood of match (more often is better)

mwm-deriv committed:
Update for CPAN release

- Changes file updated
- Makefile.PL updated by dzil

shaydo-deriv commented on LICENSE:
```diff
@@ -0,0 +1,412 @@
+This software is copyright (c) 2022 by Binary Services Ltd.
```
@mwm-deriv I see that the LICENSE is present in distribution on CPAN. I expect that this file is added by distzilla during release process, so it needs not to be added

mwm-deriv replied:
Ok.  Just doing `dzil test` produced the LICENSE, so I added it in.  Easily dropped.

shaydo-deriv commented on lib/Date/Utility.pm:
```diff
@@ -407,51 +401,34 @@ sub _build_is_a_weekday {
-
-    my $obj = $popular{$new_params->{epoch}};
+use constant EPOCH_RE         => qr/^-?[0-9]{1,13}$/o;
+use constant EPOCH_MAYBE_FRAC => qr/^-?[0-9]{1,13}(?:\.[0-9]+)?$/o;
```
where is 13 digits limit comes from? don't care, just curious.

mwm-deriv replied:
So... it is apparently acceptable to produce a "date string" like `YYYYMMDDHHMMSS` which is 14 digits.  From comments in there,  I believe it also used to be acceptable to produce `YYYYMMDD` so it was clamped at `{9,13}`.

shaydo-deriv reviewed (commented):
code changes look good to me, just question regarding if LICENSE is needed and if the AUTHOR should be BINARY or DERIV. I assume the upload is going to happen under DERIV account.

mwm-deriv commented:
> code changes look good to me, just question regarding if LICENSE is needed and if the AUTHOR should be BINARY or DERIV. I assume the upload is going to happen under DERIV account.

I looked this up on MetaCPAN and it's owned by `BINARY`.  WIki docs don't suggest moving things between which I imagine to be a big hassle.

shaydo-deriv commented:
> > code changes look good to me, just question regarding if LICENSE is needed and if the AUTHOR should be BINARY or DERIV. I assume the upload is going to happen under DERIV account.
> 
> I looked this up on MetaCPAN and it's owned by `BINARY`. WIki docs don't suggest moving things between which I imagine to be a big hassle.



> > code changes look good to me, just question regarding if LICENSE is needed and if the AUTHOR should be BINARY or DERIV. I assume the upload is going to happen under DERIV account.
> 
> I looked this up on MetaCPAN and it's owned by `BINARY`. WIki docs don't suggest moving things between which I imagine to be a big hassle.

ok, then just LICENSE file question

mwm-deriv committed:
Update dist.ini for DERIV

mwm-deriv committed:
Release Date-Utility 1.12

	Remove dependency on Tie::Hash::LRU
	Switch dependency from Moose to Moo
	Improve datatime parsing performance
